pkg/tools/ha/lease.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//This utility uses the Lock API in the client to acquire a lock.
+//It provides a uniform and consistent semantics for.
+// - Logging how many times a lease is gained/lost.
+// - Starting/Stopping a daemon (via callbacks).
+// - Logging errors which might occur if a daemon's lease isn't consistent with its running state.
+package ha
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client"
+	"k8s.io/kubernetes/pkg/client/clientcmd"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/clientcmd/api"
+	"os"
+	"time"
+)
+
+// A program that is running may "use" a lease.
+// This struct has some metadata about that program.
+// It is used for debugging/logging (i.e. print how many times we got the lease)
+// It also is used to track the state of a program (i.e. if lease obtained, Running should be true)
+type LeaseUser struct {
+
+	//Note: In general 'Running' should be 'false' when you first start an HA process, and it should
+	//be flipped to 'true' once you acquire the lock and start up.
+	Running      bool // This is how users verify that their process is running.
+	LeasesGained int  // Number of times lease has been granted.
+	LeasesLost   int  // Number of times the lease was lost.
+}
+
+// Configuration options needed for running the lease loop.
+type Config struct {
+	Key       string
+	whoami    string
+	ttl       uint64
+	sleep     time.Duration
+	lastLease time.Time
+	// These two functions return "err" or else nil.
+	// They should also update information in the LeaseUserInfo struct
+	// about the state of the lease owner.
+	LeaseGained   func(lu *LeaseUser) bool
+	LeaseLost     func(lu *LeaseUser) bool
+	Cli           *client.Client
+	LeaseUserInfo *LeaseUser
+}
+
+// RunHA runs a process in a highly available fashion.
+func RunHA(kubecfg string, master string, start func(l *LeaseUser) bool, stop func(l *LeaseUser) bool, lockName string) {
+
+	//We need a kubeconfig in order to use the locking API, so we create it here.
+	kubeconfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
```

Comment:
This is not the right place for config generation. A kubeClient should be passed in.